plugins/base/base.go

```diff
@@ -65,37 +78,130 @@ type ClientDriverConfig struct {
 			ClientMaxPort: uint32(c.Driver.ClientMaxPort),
 			ClientMinPort: uint32(c.Driver.ClientMinPort),
+			Topology:      nomadTopologyToProto(c.Driver.Topology),
 		}
 	}
-
 	return cfg
 }
 
 func nomadConfigFromProto(pb *proto.NomadConfig) *AgentConfig {
 	if pb == nil {
 		return nil
 	}
-
 	cfg := &AgentConfig{}
 	if pb.Driver != nil {
 		cfg.Driver = &ClientDriverConfig{
 			ClientMaxPort: uint(pb.Driver.ClientMaxPort),
 			ClientMinPort: uint(pb.Driver.ClientMinPort),
+			Topology:      nomadTopologyFromProto(pb.Driver.Topology),
 		}
 	}
-
 	return cfg
 }
+
+func nomadTopologyFromProto(pb *proto.ClientTopology) *numalib.Topology {
+	if pb == nil {
+		return nil
+	}
+	return &numalib.Topology{
+		NodeIDs:                idset.FromFunc(pb.NodeIds, func(i uint32) hw.NodeID { return hw.NodeID(i) }),
+		Distances:              nomadTopologyDistancesFromProto(pb.Distances),
+		Cores:                  nomadTopologyCoresFromProto(pb.Cores),
+		OverrideTotalCompute:   hw.MHz(pb.OverrideTotalCompute),
+		OverrideWitholdCompute: hw.MHz(pb.OverrideWitholdCompute),
+	}
+}
+
+func nomadTopologyDistancesFromProto(pb *proto.ClientTopologySLIT) numalib.SLIT {
+	if pb == nil {
+		return nil
+	}
+	size := int(pb.Dimension)
+	slit := make(numalib.SLIT, size)
+	for row := 0; row < size; row++ {
+		slit[row] = make([]numalib.Cost, size)
+		for col := 0; col < size; col++ {
+			index := row*size + col
+			slit[row][col] = numalib.Cost(pb.Values[index])
+		}
+	}
+	return slit
+}
+
+func nomadTopologyCoresFromProto(pb []*proto.ClientTopologyCore) []numalib.Core {
+	if len(pb) == 0 {
+		return nil
+	}
+	return helper.ConvertSlice(pb, func(pbcore *proto.ClientTopologyCore) numalib.Core {
+		return numalib.Core{
+			SocketID:   hw.SocketID(pbcore.SocketId),
+			NodeID:     hw.NodeID(pbcore.NodeId),
+			ID:         hw.CoreID(pbcore.CoreId),
+			Grade:      numalib.CoreGrade(pbcore.CoreGrade),
+			Disable:    pbcore.Disable,
+			BaseSpeed:  hw.MHz(pbcore.BaseSpeed),
+			MaxSpeed:   hw.MHz(pbcore.MaxSpeed),
+			GuessSpeed: hw.MHz(pbcore.GuessSpeed),
+		}
+	})
+}
+
+func nomadTopologyToProto(top *numalib.Topology) *proto.ClientTopology {
+	if top == nil {
+		return nil
+	}
+	return &proto.ClientTopology{
+		NodeIds:                helper.ConvertSlice(top.NodeIDs.Slice(), func(id hw.NodeID) uint32 { return uint32(id) }),
+		Distances:              nomadTopologyDistancesToProto(top.Distances),
+		Cores:                  nomadTopologyCoresToProto(top.Cores),
+		OverrideTotalCompute:   uint64(top.OverrideTotalCompute),
+		OverrideWitholdCompute: uint64(top.OverrideWitholdCompute),
+	}
+}
+
+func nomadTopologyDistancesToProto(slit numalib.SLIT) *proto.ClientTopologySLIT {
+	dimension := len(slit)
+	values := make([]uint32, 0, dimension)
+	for row := 0; row < dimension; row++ {
+		for col := 0; col < dimension; col++ {
+			values = append(values, uint32(slit[row][col]))
+		}
+	}
+	return &proto.ClientTopologySLIT{
+		Dimension: uint32(dimension),
+		Values:    values,
+	}
+}
```

Comment:
We have a single dimension on the SLIT table which implies it's always a square matrix, and I'm having trouble envisioning how that could be the case (I'm almost certainly missing something). This or the protobuf might not be a bad place to leave a tiny piece of domain knowledge in the form of a docstring about what how this table dimension works

Reply:
The SLIT table is just a way of expressing "how expensive is it for a process on one numa node to access memory on another numa node". If we have nodes [0, N] we would need a mapping from each node to [0, N], hence a square.

Reply:
Ok, gotcha. I think what was confusing me is a couple of the test cases have tables that can't exist, like:

```go
NodeIDs:   idset.From[hw.NodeID]([]hw.NodeID{0, 1}),
Distances: numalib.SLIT{{10}},
```

A table for 2 nodes could only have a dimension of 1 if the node-to-self relationship was excluded, but you can't make a square matrix [0-2] if you do that.

It's fine that test cases are implausible I suppose, maybe even good b/c it means the functions are robust against weird input. But it made for that "what am I missing here?" moment. :grinning:

Reply:
Ah good catch, I've fixed up the tests to be plausible!